### PR TITLE
ARM64: Fix GC encoding settings

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
This change fixes certain settings in the GCInfo encoder/decoders.
It also fixes the debug info for the stack pointer.

### CODE REVIEW

**Comment by swaroop-sridhar** [comment]
@dotnet-bot test Windows_NT arm64

**Comment by swaroop-sridhar** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by swaroop-sridhar** [comment]
@rahku  please review

**Comment by swaroop-sridhar** [comment]
#3825

**Comment by kyulee1** [comment]
@dotnet-bot test Windows_NT arm64 Checked

File: src/inc/gcinfotypes.h [hunk 1]
```diff
@@ -255,24 +255,22 @@ struct GcStackSlot
...
-#define NORMALIZE_STACK_SLOT(x) ((x)>>3)
+#define NORMALIZE_STACK_SLOT(x) ((x)>>3)   // GC Pointers word aligned
 #define DENORMALIZE_STACK_SLOT(x) ((x)<<3)
-#define NORMALIZE_CODE_LENGTH(x) (x)
-#define DENORMALIZE_CODE_LENGTH(x) (x)
-#define NORMALIZE_STACK_BASE_REGISTER(x) ((x) ^ 5)
-#define DENORMALIZE_STACK_BASE_REGISTER(x) ((x) ^ 5)
+#define NORMALIZE_CODE_LENGTH(x) ((x)>>2)   // All Instructions are 4 bytes long
+#define DENORMALIZE_CODE_LENGTH(x) ((x)<<2) 
+#define NORMALIZE_STACK_BASE_REGISTER(x) ((x)^29) // Encode Frame pointer X29 as zero
```
Comment (rahku) [inline src/inc/gcinfotypes.h]: can you check what we do for arm arch applies for arm64

File: src/inc/gcinfotypes.h [hunk 2]
```diff
@@ -255,24 +255,22 @@ struct GcStackSlot
...
 
-// ARM64TODO: these were copied from AMD64, and need to be reviewed
-
 #ifndef TARGET_POINTER_SIZE
 #define TARGET_POINTER_SIZE 8    // equal to sizeof(void*) and the managed pointer size in bytes for this target
 #endif
 #define NUM_NORM_CODE_OFFSETS_PER_CHUNK (64)
 #define NUM_NORM_CODE_OFFSETS_PER_CHUNK_LOG2 (6)
-#define NORMALIZE_STACK_SLOT(x) ((x)>>3)
+#define NORMALIZE_STACK_SLOT(x) ((x)>>3)   // GC Pointers word aligned
```
Comment (rahku) [inline src/inc/gcinfotypes.h]: word is 16-bits

File: src/inc/gcinfotypes.h [hunk 3]
```diff
@@ -285,10 +283,10 @@ struct GcStackSlot
 #define SECURITY_OBJECT_STACK_SLOT_ENCBASE 6
 #define GS_COOKIE_STACK_SLOT_ENCBASE 6
 #define CODE_LENGTH_ENCBASE 8
-#define STACK_BASE_REGISTER_ENCBASE 3
+#define STACK_BASE_REGISTER_ENCBASE 1 // FP encoded as 0
```
Comment (rahku) [inline src/inc/gcinfotypes.h]: should this be 2 as x30 is LR

File: src/inc/gcinfotypes.h [hunk 4]
```diff
@@ -285,10 +283,10 @@ struct GcStackSlot
...
 #define SIZE_OF_STACK_AREA_ENCBASE 3
 #define SIZE_OF_EDIT_AND_CONTINUE_PRESERVED_AREA_ENCBASE 4
-#define NUM_REGISTERS_ENCBASE 2
+#define NUM_REGISTERS_ENCBASE 3
```
Comment (rahku) [inline src/inc/gcinfotypes.h]: why is this modified to be 3? Comment in related section for arm says that this is only related to perf and not correctness

File: src/inc/gcinfotypes.h [hunk 5]
```diff
@@ -300,7 +298,7 @@ struct GcStackSlot
 #define REGISTER_DELTA_ENCBASE 2
 #define STACK_SLOT_ENCBASE 6
 #define STACK_SLOT_DELTA_ENCBASE 4
-#define NUM_SAFE_POINTS_ENCBASE 2
+#define NUM_SAFE_POINTS_ENCBASE 3
```
Comment (rahku) [inline src/inc/gcinfotypes.h]: same here as well

File: src/inc/gcinfotypes.h (hunk 1 above)
Comment (swaroop-sridhar) [inline src/inc/gcinfotypes.h]: NORMALIZE_STACK_BASE_REGISTER() encodes FP as 0 and SP as 2, to minimize the number of bits necessary to encode the base-register for reported stack slots. So, I believe this encoding is sufficient for ARM64.

File: src/inc/gcinfotypes.h (hunk 2 above)
Comment (swaroop-sridhar) [inline src/inc/gcinfotypes.h]: Fixed the comment

File: src/inc/gcinfotypes.h (hunk 4 above)
Comment (swaroop-sridhar) [inline src/inc/gcinfotypes.h]: Yes, this is not a correctness issue, but based on empirical observation -- most safepoints reported more than 4 registers.

**Comment by swaroop-sridhar** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by rahku** [comment]
Looks good

**Comment by swaroop-sridhar** [comment]
@dotnet-bot test Ubuntu x64 Checked 
@dotnet-bot test OSX x64 Checked

**Comment by swaroop-sridhar** [comment]
@dotnet-bot retest this please
